File: libs/go/identifiers/kind.go
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

package identifiers

const (
	// MinimumKindLength and MaximumKindLength bound identifier prefixes.
	MinimumKindLength = 2
	MaximumKindLength = 24
)

// Kind is the canonical lowercase prefix of a resource ID, such as "run",
// "org", or "model". Kinds contain only ASCII lowercase letters and digits,
// begin with a letter, and do not contain the underscore separator.
type Kind string

// ParseKind validates value as a canonical Kind.
func ParseKind(value string) (Kind, error) {
	kind := Kind(value)
	if err := kind.Validate(); err != nil {
		return "", err
	}
	return kind, nil
}

// MustParseKind is ParseKind for package-level constants and test fixtures. It
// panics when value is invalid and must not be used on untrusted input.
func MustParseKind(value string) Kind {
	kind, err := ParseKind(value)
	if err != nil {
		panic(err)
	}
	return kind
}

// String returns the canonical prefix.
func (kind Kind) String() string {
	return string(kind)
}

// Valid reports whether kind is canonical.
func (kind Kind) Valid() bool {
	return kind.Validate() == nil
}

// Validate checks the canonical Kind grammar.
func (kind Kind) Validate() error {
	value := string(kind)
	if len(value) < MinimumKindLength || len(value) > MaximumKindLength {
		return invalidValue(
			"kind",
			value,
			"length must be between 2 and 24 bytes",
			ErrInvalidKind,
		)
	}

	for index := 0; index < len(value); index++ {
		character := value[index]
		isLetter := character >= 'a' && character <= 'z'
		isDigit := character >= '0' && character <= '9'
		if index == 0 && !isLetter {
			return invalidValue(
				"kind",
				value,
				"first character must be an ASCII lowercase letter",
				ErrInvalidKind,
			)
		}
		if !isLetter && !isDigit {
			return invalidValue(
				"kind",
				value,
				"characters must be ASCII lowercase letters or digits",
				ErrInvalidKind,
			)
		}
	}
	return nil
}

File: libs/rust/identifiers/src/kind.rs
// Copyright © 2026 Mindclade, LLC. All Rights Reserved.
// Mindclade Proprietary and Confidential.
// SPDX-License-Identifier: LicenseRef-Mindclade-Proprietary
//

//! The validated prefix half of a resource identifier.
//!
//! There is one kind grammar and it lives in `resource_id`. This type used to
//! carry a second one — 1 to 48 bytes of `[a-z0-9]` plus `_` and `-` after the
//! first — which admitted values the identifier format cannot carry.
//! `ResourceKind::parse("runtime_host")` succeeded while
//! `ResourceId::parse("runtime_host_<32 hex>")` failed on "more than one
//! separator", and `libs/go/identifiers.ParseKind` and
//! `libs/python/identifiers.parse_kind` both reject `run_id`, `run-id`, `1run`
//! and `a` outright. A validator that accepts what the wire format rejects is a
//! parse failure waiting at the language boundary.

use crate::resource_id::validate_kind;
use mindclade_faults::{Fault, FaultResult};

#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct ResourceKind(String);

impl ResourceKind {
    pub fn parse(value: impl Into<String>) -> FaultResult<Self> {
        let value = value.into();
        validate_kind(&value).map_err(|error| {
            Fault::invalid_argument("resource kind is invalid").with_source(error)
        })?;
        Ok(Self(value))
    }
    #[must_use]
    pub fn as_str(&self) -> &str {
        &self.0
    }
}
